crates/dekaf/src/session.rs

```diff
@@ -1,33 +1,55 @@
 use super::{fetch_all_collection_names, App, Collection, Read};
+use crate::{
+    from_downstream_topic_name, from_upstream_topic_name, to_downstream_topic_name,
+    to_upstream_topic_name, ConfigOptions,
+};
 use anyhow::Context;
+use bytes::{BufMut, BytesMut};
+use gazette::broker;
 use kafka_protocol::{
-    error::ResponseError,
+    error::{ParseResponseErrorCode, ResponseError},
     indexmap::IndexMap,
-    messages::{self, metadata_response::MetadataResponseTopic, TopicName},
-    protocol::{Builder, StrBytes},
+    messages::{
+        self,
+        metadata_response::{MetadataResponsePartition, MetadataResponseTopic},
+        ConsumerProtocolAssignment, ConsumerProtocolSubscription, ListGroupsResponse,
+        RequestHeader, TopicName,
+    },
+    protocol::{buf::ByteBuf, Decodable, Encodable, StrBytes},
 };
 use std::collections::HashMap;
 use std::sync::Arc;
+use tracing::instrument;
 
 struct PendingRead {
     offset: i64,          // Journal offset to be completed by this PendingRead.
     last_write_head: i64, // Most-recent observed journal write head.
-    handle: tokio::task::JoinHandle<anyhow::Result<(Read, bytes::Bytes)>>,
+    docs_remaining: Option<usize>,
+    handle: tokio::task::JoinHandle<anyhow::Result<(Read, usize, bytes::Bytes)>>,
 }
 
 pub struct Session {
     app: Arc<App>,
     client: postgrest::Postgrest,
     reads: HashMap<(TopicName, i32), PendingRead>,
+    listed_offsets: HashMap<TopicName, HashMap<i32, Option<(i64, i64)>>>,
+    /// ID of the authenticated user
+    uid: Option<String>,
+    config: Option<ConfigOptions>,
```

Comment:
nit: I'd make it just `config: ConfigOptions` initialized with defaults, which are overwritten, and then you can avoid Some() checks. Although this doesn't appear to be accessed?

Reply:
Accessed down in `encode_topic_name()`.

AFAICT, the only reason to do that (initialize `ConfigOptions` with defaults rather than have it be `Option`al) is to avoid the runtime overhead/messyness of an `.expect()` call. Considering the non-trivial complexity involved, i.e a `Session` _is_ valid and can respond to certain API requests before authentication (like `ApiVersions`, `SaslHandshake`, etc) but not others, I'd rather have a hard failure if I try to access something that hasn't been initialized yet (such as `config`), rather than silently and subtily do the wrong thing by using a default.

The core problem is that we're representing the internal state of a state machine without being clear about it -- accessing `config` simply doesn't make any sense before a `Session` is authenticated, not to mention calling all of the APIs that rely on you being authenticated... I should probably make it explicit which APIs can be called when, shouldn't I.